datastream-utils/src/main/java/com/linkedin/datastream/common/FutureUtils.java

```diff
@@ -0,0 +1,43 @@
+/**
+ *  Copyright 2019 LinkedIn Corporation. All rights reserved.
+ *  Licensed under the BSD 2-Clause License. See the LICENSE file in the project root for license information.
+ *  See the NOTICE file in the project root for additional information regarding copyright ownership.
+ */
+package com.linkedin.datastream.common;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+
+public class FutureUtils {
+
+  private static final ScheduledExecutorService SCHEDULED_EXECUTOR_SERVICE = Executors.newScheduledThreadPool(100);
+
+  /**
+   * This function waits till timeout for Future to complete, otherwise, it cancels it.
+   * @param futureToAwait Future to complete/stop
+   * @param timeout time to wait before aborting future
+   * @return CompletableFuture for this operation
+   */
+  public static CompletableFuture<?> awaitFutureOrCancel(Future<?> futureToAwait, Duration timeout) {
```

Comment:
I know our discussions have been mostly focused on the specific use-case that motivated the introduction of this API. However, we need to consider how such a fundamental utility will be used in general.

For instance:
1. If I'm calling this API, wouldn't I naturally find it useful if it offered me a way to access the result of `futureToAwait` in case it completed successfully? This should be fairly straightforward for this API to accomplish because all it takes in this case is calling `futureToAwait.get()`, sparing the caller this effort.

2. Wouldn't it be useful if the returned `CompletableFuture` communicated the success/failure/cancellation of `futureToAwait` instead? One of the annoying things about `Future` in Java is the lack of composability; there's no straightforward way to chain other futures to it. However, this isn't the case with `CompletableFuture`. So, as a caller, I would imagine composing the returned future to react to success and cancellation/failure of the future I passed. This would allow me to do:
```java
    CompletableFuture<?> result = awaitFutureOrCancel(future, timeout);
    result
        .exceptionally()   // future didn't complete successfully (threw or got cancelled)
        .thenRun();        // future completed successfully
```

Here's a proposed impl that addresses these two issues:
```java
  public static <T> CompletableFuture<T> awaitFutureOrCancel(Future<T> future, Duration timeout) {
    CompletableFuture<T> result = new CompletableFuture<>();

    SCHEDULED_EXECUTOR_SERVICE.schedule(() -> {
      if (future.isDone()) {
        try {
          T t = future.get();
          result.complete(t);
        } catch (InterruptedException e) {
          result.completeExceptionally(e);
        } catch (ExecutionException e) {
          result.completeExceptionally(e.getCause());
        } catch (CancellationException e) {
          result.cancel(false);
        }
      } else {
        future.cancel(true);
        result.cancel(false);
      }
    }, timeout.toMillis(), TimeUnit.MILLISECONDS);
    return result;
  }
```